packages/core/content-manager/admin/src/pages/EditSettingsView/reducer.js

```diff
@@ -1,126 +1,188 @@
-import { fromJS } from 'immutable';
+import produce from 'immer';
+import set from 'lodash/set';
+import get from 'lodash/get';
+import cloneDeep from 'lodash/cloneDeep';
+
+import { arrayMoveItem } from '../../utils';
 import { formatLayout, getInputSize } from './utils/layout';
 
-const initialState = fromJS({
+const initialState = {
   fieldForm: {},
   componentLayouts: {},
   metaToEdit: '',
   initialData: {},
   metaForm: {},
   modifiedData: {},
-});
-
-const reducer = (state, action) => {
-  const layoutPathEdit = ['modifiedData', 'layouts', 'edit'];
-  const layoutPathRelations = ['modifiedData', 'layouts', 'editRelations'];
+};
 
-  switch (action.type) {
-    case 'ADD_RELATION':
-      return state.updateIn(layoutPathRelations, list => list.push(action.name));
-    case 'MOVE_RELATION': {
-      return state.updateIn(layoutPathRelations, list => {
-        return list
-          .delete(action.dragIndex)
-          .insert(action.hoverIndex, state.getIn([...layoutPathRelations, action.dragIndex]));
-      });
-    }
-    case 'MOVE_ROW':
-      return state.updateIn(layoutPathEdit, list => {
-        return list
-          .delete(action.dragRowIndex)
-          .insert(action.hoverRowIndex, state.getIn([...layoutPathEdit, action.dragRowIndex]));
-      });
-    case 'ON_ADD_DATA': {
-      const size = getInputSize(state.getIn(['modifiedData', 'attributes', action.name, 'type']));
+const reducer = (state = initialState, action) =>
+  // eslint-disable-next-line consistent-return
+  produce(state, draftState => {
+    const layoutPathEdit = ['modifiedData', 'layouts', 'edit'];
+    const layoutPathRelations = ['modifiedData', 'layouts', 'editRelations'];
+    switch (action.type) {
+      case 'ADD_RELATION': {
+        const layoutFieldList = get(state, layoutPathRelations, []);
```

Comment:
WDYT of `const editRelationLayoutValue` ?

Also the editRelations is always defined so you can just do the following:

```
draftState.modifiedData.layouts.editRelations.push(action.name);
```

Reply:
It is always defined implicitly because the state is populated by the fetched data before but not explicitly. As you can see, `modifiedData: {}` in the initial state.

Reply:
Then WDYT of adding the known properties to both the initialData and modified data then?